### PR TITLE
feat(planner): Introduce bitmap to record applied rules

### LINKED ISSUE #9536: Feature: Introduce bitmap to record applied rules
So far, we are implementing the structure `AppliedRules` with a `HashSet<RuleID>`, which is inefficient. It's possible to improve this by introducing a real bitmap data structure.

cc @dusx1981

### PR DESCRIPTION
**Pull request by dusx1981**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

A draft pr for code review.

closes #9536

### CODE REVIEW

File: src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs [hunk 1]
```diff
@@ -149,17 +153,34 @@ impl ExploreExprTask {
             .memo
             .group(self.group_index)?
             .m_expr(self.m_expr_index)?;
+        let rule_set: roaring::RoaringBitmap;
+        unsafe {
+            rule_set = m_expr
+                .plan
+                .exploration_candidate_rules()
+                .bitand(&RULE_FACTORY.exploration_rules);
+        }
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs:162]: It's better to wrap the unsafe code with a safe function call.

Comment (dusx1981) [reply src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs:162]: ok, I'll resolve it tonight.

**Comment by dusx1981** [comment]
[code to be reviewed](https://github.com/datafuselabs/databend/commit/ace2d3758918d274eb2f96fe9a6702b3f244d15e)

File: src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs [hunk 2]
```diff
@@ -149,17 +162,28 @@ impl ExploreExprTask {
...
+        // for rule in optimizer.explore_rules.iter() {
+        //     let apply_rule_task = ApplyRuleTask::with_parent(
+        //         rule.id(),
+        //         m_expr.group_index,
+        //         m_expr.index,
+        //         &self.ref_count,
+        //     );
+        //     scheduler.add_task(Task::ApplyRule(apply_rule_task));
+        // }
+
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs:186]: Need to remove these codes.

File: src/query/sql/src/planner/optimizer/heuristic/heuristic.rs [hunk 3]
```diff
@@ -136,6 +161,19 @@ impl HeuristicOptimizer {
...
+        //         s_expr.set_applied_rule(&rule.id());
+        //         rule.apply(&s_expr, &mut state)?;
+        //         if !state.results().is_empty() {
+        //             // Recursive optimize the result
+        //             let result = &state.results()[0];
+        //             let optimized_result = self.optimize_expression(result)?;
+        //             return Ok(optimized_result);
+        //         }
+        //     }
+        // }
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/heuristic/heuristic.rs:176]: Need to remove these codes.

File: src/query/sql/src/planner/optimizer/rule/factory.rs [hunk 4]
```diff
@@ -44,11 +47,48 @@ use crate::optimizer::rule::RuleID;
...
+        // }
+
+        // for rule in self.rule_set.iter() {
+        //     if rule.transformation() {
+        //         self.transformation_rules.insert(rule.id() as u32);
+        //     } else {
+        //         self.exploration_rules.insert(rule.id() as u32);
+        //     }
+        // }
+    }
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/rule/factory.rs:88]: Is this part still working in progress?

**Comment by leiysky** [comment]
The function `calc_operator_rule_set` looks good to me.

**Comment by dusx1981** [comment]
Thank you for your affirmation. :)

**Comment by dusx1981** [comment]
Because work on weekdays, I can only submit codes at night or on weekends. If the feedback is not timely, please forgive me.

**Comment by dusx1981** [comment]
Is there still a problem with the code, can this PR be merged？

File: src/query/sql/src/planner/optimizer/rule/factory.rs [hunk 5]
```diff
@@ -44,11 +47,30 @@ use crate::optimizer::rule::RuleID;
...
+
+    pub transformation_rules: roaring::RoaringBitmap,
+    pub exploration_rules: roaring::RoaringBitmap,
+}
 
 impl RuleFactory {
     pub fn create() -> Self {
-        RuleFactory {}
+        RuleFactory {
+            rule_set: RuleSet::create(),
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/rule/factory.rs:63]: Maybe this should be initialized manually with all `RuleID`s filled?

**Comment by leiysky** [comment]
> Is there still a problem with the code, can this PR be merged？ could i join the follow-up development of the optimizer?

Yes, you may need to run `make lint` to check your code on your own machine, and then run the tests with github actions.

**Comment by dusx1981** [comment]
I see many other commits where the automation fails. How can I tell if there is a problem with automation?

**Comment by dusx1981** [comment]
make lint err: **error when handling /home/dusx/databend/src/query/pipeline/transforms/Cargo.toml: missing field `ignored` for key `package.metadata.cargo-machete` at line 20 column 1**, But I haven't changed this file.

**Comment by dusx1981** [comment]
make lint err: error when handling /home/dusx/databend/src/query/pipeline/transforms/Cargo.toml: missing field ignored for key package.metadata.cargo-machete at line 20 column 1, But I haven't changed this file.

**Comment by BohuTANG** [comment]
Some tests are broken, like(You can check them with details url):
https://github.com/datafuselabs/databend/actions/runs/4204816701/jobs/7296210266#step:4:390

**Comment by Xuanwo** [comment]
> Rule: CLA Check (post_check)

**request**: Please attact CLA as requried: 

```markdown
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Close #issue
```

> Rule: Check PR conflicts (post_check)

**request**: Please mark sure PR doesn't have conflicts so that we can run our tests.

File: src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs [hunk 6]
```diff
@@ -22,6 +23,10 @@ use super::Task;
 use crate::optimizer::cascades::scheduler::Scheduler;
 use crate::optimizer::cascades::tasks::SharedCounter;
 use crate::optimizer::cascades::CascadesOptimizer;
+// use crate::optimizer::RuleID;
```
Comment (Xuanwo) [inline src/query/sql/src/planner/optimizer/cascades/tasks/explore_expr.rs:26]: **request**: Please remove unused code.

File: src/query/sql/src/planner/optimizer/heuristic/heuristic.rs [hunk 7]
```diff
@@ -119,11 +124,32 @@ impl HeuristicOptimizer {
         Ok(result)
     }
 
+    fn calc_operator_rule_set(&self, operator: &RelOperator) -> roaring::RoaringBitmap {
+        unsafe {
+            operator
+                .transrormation_candidate_rules()
+                .bitand(&RULE_FACTORY.transformation_rules)
```
Comment (Xuanwo) [inline src/query/sql/src/planner/optimizer/heuristic/heuristic.rs:131]: **suggestion**: it's better to use `&` instead.

File: src/query/sql/src/planner/optimizer/heuristic/heuristic.rs [hunk 8]
```diff
@@ -119,11 +124,32 @@ impl HeuristicOptimizer {
...
+                DEFAULT_REWRITE_RULES[rule_id as usize],
+                Some(self.metadata.clone()),
+            )
+        }
+    }
+
     /// Try to apply the rules to the expression.
     /// Return the final result that no rule can be applied.
-    fn apply_transform_rules(&self, s_expr: &SExpr, rule_list: &RuleList) -> Result<SExpr> {
+    fn apply_transform_rules(&self, s_expr: &SExpr, _rule_list: &RuleList) -> Result<SExpr> {
```
Comment (Xuanwo) [inline src/query/sql/src/planner/optimizer/heuristic/heuristic.rs:147]: **suggestion**: It's better to remove this arg if not needed.

**Comment by dusx1981** [comment]
I use the old code to do comparative coverage test, the test of the same branch a few days ago，Error with only one comment：
image

After merging branches yesterday，A timeout error occurred:
image

**Is this a problem?**

**Comment by dusx1981** [comment]
### I was troubleshooting automated testing errors and found a problem:

> The results of HeuristicOptimizer optimizations depend on the order in which the rules are applied. In order to do a comparison test to solve this problem, I keep the old code, and after solving the problem, the old code will be cleaned up.

**Comment by Xuanwo** [comment]
> image

This issue caused by you didn't use our PR description template, and now it has been fixed.

**Comment by dusx1981** [comment]
image

**Is this timeout a problem?**

**Comment by dusx1981** [comment]
> 

I understand the problem you mentioned, and I want to modify it after solving the program problem:)

**Comment by dusx1981** [comment]
### The problem with rule application has been solved and the code can already be reviewed.

**Comment by dusx1981** [comment]
**Solve the problem:**

1. (x..y) is a half-open and half-closed interval, the initial implementation misses the last rule
2. The optimization result of HeuristicOptimizer is related to the order of the application rules, and the adjustment of the application rules is consistent with the original one.

File: src/query/sql/src/planner/optimizer/heuristic/rule_list.rs [hunk 9]
```diff
@@ -26,16 +21,16 @@ pub struct RuleList {
...
+    //     }
+    //     Ok(RuleList { rules })
+    // }
 
-    pub fn iter(&self) -> impl Iterator<Item = &RulePtr> {
-        self.rules.iter()
-    }
+    // pub fn iter(&self) -> impl Iterator<Item = &RulePtr> {
+    //     self.rules.iter()
+    // }
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/heuristic/rule_list.rs:35]: No need to modify these?

**Review by leiysky — APPROVED** [review]
Rest LGTM

**Comment by dusx1981** [comment]
> > Rule: CLA Check (post_check)
> 
> **request**: Please attach CLA as requried:
> 
> ```gfm
> I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/
> 
> ## Summary
> 
> Summary about this PR
> 
> Close #issue
> ```
> 
> > Rule: Check PR conflicts (post_check)
> 
> **request**: Please mark sure PR doesn't have conflicts so that we can run our tests.

looking forward to your review :).

**Comment by dusx1981** [comment]
> > image
> 
> This issue caused by you didn't use our PR description template, and now it has been fixed.

looking forward to your review :).

**Comment by xudong963** [comment]
Just notice the ticket, let me take a look before merging...

**Comment by BohuTANG** [comment]
Hi @dusx1981 ,

Nice, thanks for your contribution!
Can you give more summary for this PR? It would be great to add some logic tests for it if possible(Like https://github.com/datafuselabs/databend/blob/d5ae263b39dd8cf3a17d6be58b6cd1aa2439fafa/tests/sqllogictests/suites/mode/standalone/explain/limit.test#L2).

**Comment by xudong963** [comment]
> It would be great to add some logic tests for it if possible

I think the current tests are covered. The ticket only changes the way to record optimization rules.

File: src/query/sql/src/planner/optimizer/cascades/cascade.rs [hunk 10]
```diff
@@ -32,34 +34,44 @@ use crate::optimizer::rule::RuleSet;
 use crate::optimizer::rule::TransformResult;
 use crate::optimizer::SExpr;
 use crate::IndexType;
+use crate::MetadataRef;
 
 /// A cascades-style search engine to enumerate possible alternations of a relational expression and
 /// find the optimal one.
 pub struct CascadesOptimizer {
     pub memo: Memo,
     pub explore_rules: RuleSet,
```
Comment (xudong963) [inline src/query/sql/src/planner/optimizer/cascades/cascade.rs:43]: The var is useless?

Comment (xudong963) [reply src/query/sql/src/planner/optimizer/cascades/cascade.rs:43]: Any update? @dusx1981

**Comment by dusx1981** [comment]
Do I need to change the status to resolved after answering the question?

**Comment by dusx1981** [comment]
Metadata information in CascadesOptimizer and Memo has been removed:)

**Review by sundy-li — APPROVED** [review]

**Comment by sundy-li** [comment]
LGTM, thanks for the contribution!

**Pull request merged**